File: frontend/src/apps/translate/service.ts
/**
 * Translate Service - 翻译服务
 *
 * 后端路由: /api/v1/translate/...
 * 使用 LibreTranslate 本地离线翻译
 */

import { api } from "$shared/services/api";

// ==================== 类型定义 ====================

export interface Language {
  code: string;
  name: string;
}

export interface TranslateRequest {
  text: string;
  source?: string; // 源语言，空则自动检测
  target: string;  // 目标语言
}

export interface TranslateResponse {
  translatedText: string;
  detectedLang?: string;
}

export interface DetectRequest {
  text: string;
}

export interface DetectResponse {
  language: string;
  confidence: number;
}

export type ServicePhase =
  | "ready"
  | "starting"
  | "ready_to_start"
  | "missing"
  | "error"
  | string;

export interface ServiceStatus {
  available: boolean;
  url: string;
  message?: string;
  phase?: ServicePhase;
  image?: string;
  container?: string;
  containerRunning?: boolean;
  imageReady?: boolean;
  offlineReady?: boolean;
}

export interface EnsureResult {
  status: "success" | "starting" | "failed" | string;
  available: boolean;
  phase?: ServicePhase;
  message?: string;
}

export interface TranslateConfig {
  defaultSource: string;
  defaultTarget: string;
  serviceUrl: string;
}

// ==================== API 调用 ====================

/**
 * 翻译文本
 */
export async function translateText(req: TranslateRequest): Promise<TranslateResponse> {
  const response = await api.post<TranslateResponse>("/translate/text", req);
  return response;
}

/**
 * 检测语言
 */
export async function detectLanguage(text: string): Promise<DetectResponse> {
  const response = await api.post<DetectResponse>("/translate/detect", { text });
  return response;
}

/**
 * 获取支持的语言列表
 */
export async function getLanguages(): Promise<Language[]> {
  const response = await api.get<Language[]>("/translate/languages");
  return response;
}

/**
 * 获取服务状态
 */
export async function getStatus(): Promise<ServiceStatus> {
  const response = await api.get<ServiceStatus>("/translate/status");
  return response;
}

/**
 * 确保翻译服务已启动（复用容器 / 离线镜像，不走 Docker Store 安装）
 * 旧后端无 /ensure 时降级为轮询 /status，避免 UI 只显示 "not found"
 */
export async function ensureService(waitSec = 20): Promise<EnsureResult> {
  try {
    return await api.post<EnsureResult>(`/translate/ensure?wait=${waitSec}`);
  } catch (e: unknown) {
    const status = (e as { status?: number })?.status;
    const message = (e as { message?: string })?.message || "";
    if (status !== 404 && !/not found/i.test(message)) {
      throw e;
    }
    // 兼容旧包：无 ensure 路由时轮询 status
    const deadline = Date.now() + Math.max(waitSec, 5) * 1000;
    while (Date.now() < deadline) {
      const st = await getStatus().catch(() => null);
      if (st?.available) {
        return {
          status: "success",
          available: true,
          phase: "ready",
          message: st.message || "翻译服务已就绪",
        };
      }
      await new Promise((r) => setTimeout(r, 2000));
    }
    const st = await getStatus().catch(() => null);
    if (st?.available) {
      return { status: "success", available: true, phase: "ready", message: st.message };
    }
    return {
      status: "failed",
      available: false,
      phase: st?.phase || "error",
      message:
        st?.message ||
        "当前系统版本缺少翻译启动接口，请升级 RDE 或等待离线引导完成后再试",
    };
  }
}

/**
 * 获取翻译配置
 * @param lang 系统语言（用于确定默认翻译方向）
 */
export async function getConfig(lang?: string): Promise<TranslateConfig> {
  const params = lang ? `?lang=${lang}` : "";
  const response = await api.get<TranslateConfig>(`/translate/config${params}`);
  return response;
}

// ==================== 辅助函数 ====================

/**
 * 常用语言映射（用于 UI 显示）
 */
export const COMMON_LANGUAGES: Record<string, string> = {
  en: "English",
  zh: "中文",
  ja: "日本語",
  ko: "한국어",
  fr: "Français",
  de: "Deutsch",
  es: "Español",
  pt: "Português",
  ru: "Русский",
  ar: "العربية",
  it: "Italiano",
};

/**
 * 获取语言显示名称
 */
export function getLanguageName(code: string, languages: Language[]): string {
  const lang = languages.find((l) => l.code === code);
  return lang?.name || COMMON_LANGUAGES[code] || code;
}

File: backend/modules/translate/service.go
// Package translate 翻译服务
package translate

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"

	"go.uber.org/zap"
)

const (
	// DefaultLibreTranslateURL LibreTranslate 默认地址
	DefaultLibreTranslateURL = "http://localhost:5000"

	// 默认超时时间
	defaultTimeout = 30 * time.Second
)

// Service 翻译服务
type Service struct {
	logger     *zap.Logger
	client     *http.Client
	serviceURL string
	mu         sync.RWMutex

	// 缓存语言列表
	languages     []Language
	languagesTime time.Time
}

// NewService 创建翻译服务
func NewService(logger *zap.Logger) *Service {
	return &Service{
		logger:     logger,
		serviceURL: DefaultLibreTranslateURL,
		client: &http.Client{
			Timeout: defaultTimeout,
		},
	}
}

// SetServiceURL 设置服务地址
func (s *Service) SetServiceURL(url string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.serviceURL = url
}

// GetServiceURL 获取服务地址
func (s *Service) GetServiceURL() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.serviceURL
}

// Translate 翻译文本
func (s *Service) Translate(req TranslateRequest) (*TranslateResponse, error) {
	s.mu.RLock()
	serviceURL := s.serviceURL
	s.mu.RUnlock()

	// 构建 LibreTranslate 请求
	libreReq := libreTranslateRequest{
		Q:      req.Text,
		Source: req.Source,
		Target: req.Target,
		Format: "text",
	}

	// 如果未指定源语言，使用自动检测
	if libreReq.Source == "" {
		libreReq.Source = "auto"
	}

	data, err := json.Marshal(libreReq)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// 发送请求
	url := serviceURL + "/translate"
	httpReq, err := http.NewRequest("POST", url, bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("translate failed with status %d: %s", resp.StatusCode, string(body))
	}

	// 解析响应
	var libreResp libreTranslateResponse
	if err := json.NewDecoder(resp.Body).Decode(&libreResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	result := &TranslateResponse{
		TranslatedText: libreResp.TranslatedText,
	}

	// 如果有检测到的语言信息
	if libreResp.DetectedLanguage != nil {
		result.DetectedLang = libreResp.DetectedLanguage.Language
	}

	return result, nil
}

// DetectLanguage 检测语言
func (s *Service) DetectLanguage(text string) (*DetectResponse, error) {
	s.mu.RLock()
	serviceURL := s.serviceURL
	s.mu.RUnlock()

	reqBody := map[string]string{"q": text}
	data, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	url := serviceURL + "/detect"
	httpReq, err := http.NewRequest("POST", url, bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("detect failed with status %d: %s", resp.StatusCode, string(body))
	}

	var libreResp libreDetectResponse
	if err := json.NewDecoder(resp.Body).Decode(&libreResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if len(libreResp) == 0 {
		return nil, fmt.Errorf("no language detected")
	}

	return &DetectResponse{
		Language:   libreResp[0].Language,
		Confidence: libreResp[0].Confidence,
	}, nil
}

// GetLanguages 获取支持的语言列表
func (s *Service) GetLanguages() ([]Language, error) {
	s.mu.RLock()
	// 如果缓存有效（5分钟内），直接返回
	if len(s.languages) > 0 && time.Since(s.languagesTime) < 5*time.Minute {
		languages := s.languages
		s.mu.RUnlock()
		return languages, nil
	}
	serviceURL := s.serviceURL
	s.mu.RUnlock()

	url := serviceURL + "/languages"
	resp, err := s.client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to get languages: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("get languages failed with status %d: %s", resp.StatusCode, string(body))
	}

	var libreLanguages []libreLanguage
	if err := json.NewDecoder(resp.Body).Decode(&libreLanguages); err != nil {
		return nil, fmt.Errorf("failed to decode languages: %w", err)
	}

	languages := make([]Language, len(libreLanguages))
	for i, lang := range libreLanguages {
		languages[i] = Language{
			Code: lang.Code,
			Name: lang.Name,
		}
	}

	// 更新缓存
	s.mu.Lock()
	s.languages = languages
	s.languagesTime = time.Now()
	s.mu.Unlock()

	return languages, nil
}

// CheckStatus 检查服务状态
func (s *Service) CheckStatus() *ServiceStatus {
	s.mu.RLock()
	serviceURL := s.serviceURL
	s.mu.RUnlock()

	status := &ServiceStatus{
		Available: false,
		URL:       serviceURL,
	}

	// 尝试获取语言列表来验证服务
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(serviceURL + "/languages")
	if err != nil {
		status.Message = fmt.Sprintf("无法连接到翻译服务: %v", err)
		return status
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		status.Message = fmt.Sprintf("翻译服务返回错误: %d", resp.StatusCode)
		return status
	}

	status.Available = true
	status.Message = "翻译服务运行正常"
	return status
}

// GetConfig 获取翻译配置（根据系统语言设置默认值）
func (s *Service) GetConfig(systemLang string) *TranslateConfig {
	config := &TranslateConfig{
		ServiceURL: s.GetServiceURL(),
	}

	// 根据系统语言设置默认翻译方向
	// 中文系统：中文 -> 英文
	// 英文系统：英文 -> 中文
	if systemLang == "zh-CN" || systemLang == "zh" {
		config.DefaultSource = "zh"
		config.DefaultTarget = "en"
	} else {
		config.DefaultSource = "en"
		config.DefaultTarget = "zh"
	}

	return config
}
